packages/excalidraw/element/distance.ts

```diff
@@ -0,0 +1,234 @@
+import type { GlobalPoint, Radians } from "../../math";
+import {
+  curve,
+  curvePointDistance,
+  distanceToLineSegment,
+  lineSegment,
+  pointFrom,
+  pointRotateRads,
+  rectangle,
+} from "../../math";
+import { ellipse, ellipseDistanceFromPoint } from "../../math/ellipse";
+import { getCornerRadius } from "../shapes";
+import { getDiamondPoints } from "./bounds";
+import type {
+  ExcalidrawBindableElement,
+  ExcalidrawDiamondElement,
+  ExcalidrawEllipseElement,
+  ExcalidrawRectanguloidElement,
+} from "./types";
+
+export const distanceToBindableElement = (
+  element: ExcalidrawBindableElement,
+  p: GlobalPoint,
+): number => {
+  switch (element.type) {
+    case "rectangle":
+    case "image":
+    case "text":
+    case "iframe":
+    case "embeddable":
+    case "frame":
+    case "magicframe":
+      return distanceToRectanguloidElement(element, p);
+    case "diamond":
+      return distanceToDiamondElement(element, p);
+    case "ellipse":
+      return distanceToEllipseElement(element, p);
+  }
+};
+
+/**
+ * Returns the distance of a point and the provided rectangular-shaped element,
+ * accounting for roundness and rotation
+ *
+ * @param element The rectanguloid element
+ * @param p The point to consider
+ * @returns The eucledian distance to the outline of the rectanguloid element
+ */
+export const distanceToRectanguloidElement = (
+  element: ExcalidrawRectanguloidElement,
+  p: GlobalPoint,
+) => {
+  const r = rectangle(
+    pointFrom(element.x, element.y),
+    pointFrom(element.x + element.width, element.y + element.height),
+  );
+  // To emulate a rotated rectangle we rotate the point in the inverse angle
+  // instead. It's all the same distance-wise.
+  const rotatedPoint = pointRotateRads(
+    p,
+    pointFrom(element.x + element.width / 2, element.y + element.height / 2),
+    -element.angle as Radians,
+  );
+  const roundness = getCornerRadius(
+    Math.min(element.width, element.height),
+    element,
+  );
+  const top = lineSegment<GlobalPoint>(
```

Comment:
If we're constructing the line segments and curves in each of the util helpers (`distanceTo...`, `intersect...`), what do you think about the following idea:
- include an (optional) threshold in the function parameter
- as each line or corner curve is constructed, calculate against the threshold

If the optional threshold is not included, we could then calculate for all and return a single value(s).

If this doesn't apply or is undesirable, maybe we could combine the line segments and corner curves retrieval in a separate helper so that the work is not repeated in different locations.

Reply:
Is this for optimization purposes?

Reply:
That makes sense. If it's for min distance, then you'd have to consider all segments. What about the putting the segments and curves retrieval for a given shape in a separate helper? I see that they are repeated for both `distanceTo...`s and `intersect...`s.